### PR TITLE
GH-40932: [Java] Implement TransferPair functionality for StringView

### PR DESCRIPTION
**Pull request by vibhatha**
### Rationale for this change

StringView implementation requires transferPair functionality which is required for C Data interface implementation as well. 

### What changes are included in this PR?

Adding transferPair functionality and corresponding test cases. 

### Are these changes tested?

Yes

### Are there any user-facing changes?

No
* GitHub Issue: apache/arrow#40932
* GitHub Issue: #40932

### CODE REVIEW

File: java/vector/src/main/java/org/apache/arrow/vector/BaseVariableWidthViewVector.java [hunk 1]
```diff
@@ -826,7 +839,120 @@ public void transferTo(BaseVariableWidthViewVector target) {
...
+      if (stringLength > INLINE_SIZE) {
+        final int bufIndex = viewBuffer.getInt(((long) i * ELEMENT_SIZE) +
+            LENGTH_WIDTH + PREFIX_WIDTH);
+        final int bufOffset = viewBuffer.getInt(((long) i * ELEMENT_SIZE) +
+            LENGTH_WIDTH + PREFIX_WIDTH + BUF_INDEX_WIDTH);
+        final ArrowBuf dataBuf = dataBuffers.get(bufIndex);
+        final ArrowBuf slicedDataBuffer = dataBuf.slice(bufOffset, stringLength);
+        ArrowBuf currentDataBuf = target.allocateOrGetLastDataBuffer(stringLength);
+        currentDataBuf.setBytes(currentDataBuf.writerIndex(), slicedDataBuffer, 0, stringLength);
+      }
```
Comment (lidavidm) [inline java/vector/src/main/java/org/apache/arrow/vector/BaseVariableWidthViewVector.java:954]: How does this work when we're not rewriting the view buffers up above?

File: java/vector/src/main/java/org/apache/arrow/vector/BaseVariableWidthViewVector.java [hunk 2]
```diff
@@ -826,7 +839,120 @@ public void transferTo(BaseVariableWidthViewVector target) {
...
+      // slice
+      if (target.validityBuffer != null) {
+        target.validityBuffer.getReferenceManager().release();
+      }
+      final ArrowBuf slicedValidityBuffer = validityBuffer.slice(firstByteSource, byteSizeTarget);
+      target.validityBuffer = transferBuffer(slicedValidityBuffer, target.allocator);
+      return;
+    }
+
+    /* Copy data
```
Comment (lidavidm) [inline java/vector/src/main/java/org/apache/arrow/vector/BaseVariableWidthViewVector.java:887]: There's not an existing helper for this? I don't think every vector should have its own copy of this code...

File: java/vector/src/test/java/org/apache/arrow/vector/TestSplitAndTransfer.java [hunk 3]
```diff
@@ -209,6 +225,39 @@ public void test() throws Exception {
     }
   }
 
+  @Test /* ViewVarCharVector */
+  public void testView() throws Exception {
```
Comment (lidavidm) [inline java/vector/src/test/java/org/apache/arrow/vector/TestSplitAndTransfer.java:229]: Instead of the comment, just rename the test?

Comment (vibhatha) [reply java/vector/src/test/java/org/apache/arrow/vector/TestSplitAndTransfer.java:229]: Ah yes, I should have got rid of that comment, already renamed to views.

File: java/vector/src/main/java/org/apache/arrow/vector/BaseVariableWidthViewVector.java (hunk 2 above)
Comment (vibhatha) [reply java/vector/src/main/java/org/apache/arrow/vector/BaseVariableWidthViewVector.java:887]: No there is no helper for this. I looked into what is `LargeListVector` and `BaseVariableWidthVector` the code is mostly the same for validity buffer related split and transfer.

Comment (vibhatha) [reply java/vector/src/main/java/org/apache/arrow/vector/BaseVariableWidthViewVector.java:887]: Shall we file a separate issue to write a util for this?

Comment (lidavidm) [reply java/vector/src/main/java/org/apache/arrow/vector/BaseVariableWidthViewVector.java:887]: That's rather unfortunate. It should go into something like BitVectorHelper

Comment (vibhatha) [reply java/vector/src/main/java/org/apache/arrow/vector/BaseVariableWidthViewVector.java:887]: I agree. Shall I file an issue and work on it as a follow up?

**Comment by vibhatha** [comment]
@lidavidm I updated the PR.

File: java/vector/src/main/java/org/apache/arrow/vector/BaseVariableWidthViewVector.java (hunk 2 above)
Comment (lidavidm) [reply java/vector/src/main/java/org/apache/arrow/vector/BaseVariableWidthViewVector.java:886]: Yes please

Comment (vibhatha) [reply java/vector/src/main/java/org/apache/arrow/vector/BaseVariableWidthViewVector.java:886]: https://github.com/apache/arrow/issues/41897

Done!

File: java/vector/src/main/java/org/apache/arrow/vector/BaseVariableWidthViewVector.java [hunk 4]
```diff
@@ -826,7 +839,158 @@ public void transferTo(BaseVariableWidthViewVector target) {
...
+      final int stringLength = getValueLength(i);
+
+      // keeping track of writing index in the target view buffer
+      int writePosition = (i - startIndex) * ELEMENT_SIZE;
+      // keeping track of reading index in the source view buffer
+      int readPosition = i * ELEMENT_SIZE;
+
+      // to clear the memory segment of view being written to
+      // this is helpful in case of overwriting the value
+      target.viewBuffer.setZero(writePosition, ELEMENT_SIZE);
```
Comment (lidavidm) [inline java/vector/src/main/java/org/apache/arrow/vector/BaseVariableWidthViewVector.java:956]: this should be unnecessary

**Review by lidavidm — APPROVED** [review]

File: java/vector/src/main/java/org/apache/arrow/vector/BaseVariableWidthViewVector.java (hunk 4 above)
Comment (vibhatha) [reply java/vector/src/main/java/org/apache/arrow/vector/BaseVariableWidthViewVector.java]: removed

**Pull request merged**